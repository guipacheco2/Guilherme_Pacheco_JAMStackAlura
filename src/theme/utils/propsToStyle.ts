import { css, FlattenSimpleInterpolation } from 'styled-components'
import { breakpointsMedia } from './breakpointsMedia'
import { createBreakpoints, ResponsiveBreakpoints } from './createBreakpoint'

export function propsToStyle(
  props: Record<string, ResponsiveBreakpoints<string | number>>,
): FlattenSimpleInterpolation[] {
  const entries = Object.entries(props)

  if (entries.length === 0) {
    throw new Error('Utilize com pelo menos uma prop')
  }

  const filledEntries = entries.filter((entry) => entry[1])

  if (filledEntries.length === 0) {
    return []
  }

  const styles = filledEntries.map(([name, value]) => {
    if (typeof value === 'string' || typeof value === 'number') {
      return css({ [name]: value })
    }

    const breakpoints = createBreakpoints(value, (v) => css({ [name]: v }))

    return breakpointsMedia(breakpoints)
  })

  return styles
}
