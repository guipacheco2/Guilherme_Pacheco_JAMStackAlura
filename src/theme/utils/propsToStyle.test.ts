import { propsToStyle } from './propsToStyle'

describe('propsToStyle()', () => {
  describe('when receives nothing', () => {
    test('throws an error', () => {
      expect(() => propsToStyle({})).toThrow('Utilize com pelo menos uma prop')
    })
  })

  describe('when receives an simple argument', () => {
    test('and it is a string', () => {
      const propsToStyleResult = propsToStyle({ textAlign: 'center' })
      expect(propsToStyleResult).toEqual([['text-align: center;']])
    })

    test('and it is a number', () => {
      const propsToStyleResult = propsToStyle({ flex: 1 })
      expect(propsToStyleResult).toEqual([['flex: 1;']])
    })
  })

  describe('when receives an argument with breakpoints', () => {
    test('renders only one breakpoint resolution', () => {
      const propsToStyleResult = propsToStyle({ textAlign: { xs: 'center' } })

      expect(propsToStyleResult).toMatchSnapshot()
    })

    test('renders two or more breakpoint resolutions', () => {
      const propsToStyleResult = propsToStyle({
        textAlign: { md: 'right', xs: 'center' },
      })

      expect(propsToStyleResult).toMatchSnapshot()
    })
  })
})
