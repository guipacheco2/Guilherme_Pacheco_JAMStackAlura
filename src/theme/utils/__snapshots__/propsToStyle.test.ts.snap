// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`propsToStyle() when receives an argument with breakpoints renders only one breakpoint resolution 1`] = `
Array [
  Array [
    Array [
      "@media only screen and (min-width:",
      "0",
      "px){",
      "text-align: center;",
      "}",
    ],
  ],
]
`;

exports[`propsToStyle() when receives an argument with breakpoints renders two or more breakpoint resolutions 1`] = `
Array [
  Array [
    Array [
      "@media only screen and (min-width:",
      "0",
      "px){",
      "text-align: center;",
      "}",
    ],
    Array [
      "@media only screen and (min-width:",
      "1024",
      "px){",
      "text-align: right;",
      "}",
    ],
  ],
]
`;
